src/main/java/edu/byu/cs/autograder/compile/StudentCodeReader.java

```diff
@@ -0,0 +1,50 @@
+package edu.byu.cs.autograder.compile;
+
+import edu.byu.cs.autograder.GradingContext;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.*;
+import java.util.stream.Stream;
+
+public class StudentCodeReader {
+    private final Set<File> files;
+    private final Map<File, List<String>> fileContents = new HashMap<>();
+
+    public static StudentCodeReader from(GradingContext context) throws IOException {
+        return new StudentCodeReader(moduleFiles(context.stageRepo(), "shared", "server", "client"));
+    }
+
+    private static Set<File> moduleFiles(File stageRepo, String... modules) throws IOException {
+        Set<File> files = new HashSet<>();
+        for(String module: modules) {
+            Path moduleRoot = Path.of(stageRepo.getPath(), module);
+            try (Stream<Path> paths = Files.walk(moduleRoot)) {
+                paths.filter(path -> path.toFile().isFile()).forEach((path -> files.add(path.toFile())));
+            }
+        }
+        return files;
+    }
+
+    private StudentCodeReader(Set<File> files) {
+        this.files = files;
+    }
+
+    public List<String> getFileContents(File file) throws IOException {
+        if(!fileContents.containsKey(file)) {
```

Comment:
I like this caching of the file contents!

After reading in the entire repo, this would be holding a lot of data in memory. After the checks have stopped, is there a way for us to manually free the resources so that they aren't preserved through the whole grading process? I assume they're freed up at the end of the student's grading run, but this seems like something we would want to free up earlier if it's no longer needed.

Reply:
Pretty sure the reader goes out of scope when the verification, modification, and compiling is over.

Reply:
But good point, It would probably be better to make it go out of scope even before we start compiling

Reply:
Now it goes out of scope after validation, before we even start modifying anything